verification/src/http_server/handlers/verification/solidity/handlers.rs

```diff
@@ -0,0 +1,106 @@
+use crate::{
+    compiler::{CompilerVersion, Compilers, CompilersError, Fetcher},
+    solidity::{VerificationSuccess, Verifier, VerifierInitializationError},
+    VerificationResponse, VerificationResult,
+};
+use actix_web::error;
+use ethers_solc::{
+    artifacts::{BytecodeHash, SettingsMetadata},
+    CompilerInput,
+};
+use std::{
+    fmt::{Debug, Display},
+    str::FromStr,
+};
+use thiserror::Error;
+
+const BYTECODE_HASHES: [BytecodeHash; 3] =
+    [BytecodeHash::Ipfs, BytecodeHash::None, BytecodeHash::Bzzr1];
+
+pub struct CompileAndVerifyInput<'a> {
+    pub compiler_version: CompilerVersion,
+    pub compiler_input: CompilerInput,
+    pub creation_tx_input: &'a str,
+    pub deployed_bytecode: &'a str,
+}
+
+#[derive(Error, Debug)]
+enum CompileAndVerifyError {
+    #[error("{0:#}")]
+    VerifierInitialization(#[from] VerifierInitializationError),
+    #[error("{0:#}")]
+    Compilation(#[from] CompilersError),
+    #[error("No contract could be verified with provided data")]
+    NoMatchingContracts,
+}
+
+pub(crate) async fn compile_and_verify_handler<T: Fetcher>(
+    compilers: &Compilers<T>,
+    mut input: CompileAndVerifyInput<'_>,
+    bruteforce_bytecode_hashes: bool,
+) -> Result<VerificationResponse, actix_web::Error>
+where
+    <T as Fetcher>::Error: Debug + Display,
+{
+    let bruteforce_metadata = if !bruteforce_bytecode_hashes {
+        Vec::from([input.compiler_input.settings.metadata.clone()])
+    } else if pre_v0_6_0(&input.compiler_version) {
+        Vec::from([None])
+    } else {
+        Vec::from(BYTECODE_HASHES.map(|hash| Some(SettingsMetadata::from(hash))))
+    };
+
+    for metadata in bruteforce_metadata {
+        input.compiler_input.settings.metadata = metadata;
+        match compile_and_verify(compilers, &input).await {
```

Comment:
can't we implement `from CompileAndVerifyError into actix_web::Error` and do something like
```rust
match compile_and_verify(compilers, &input).await {
  Ok(..)..,
  Err(NoMatchingContracts) => {},
  Err(err) => err.into(),
}
```

Reply:
In that case we would have to put some `actix_web::Error` in correspondence to `CompileAndVerifyError::NoMatchingContracts` which actually does not correspond to any actix_web error. So, I prefer to make this conversion explicit